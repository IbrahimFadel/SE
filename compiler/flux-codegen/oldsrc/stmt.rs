// use super::*;

// impl<'a> Codegen<'a> {
// 	pub fn block_stmt(&mut self, block: &BlockStmt) -> Option<FluxError> {
// 		for stmt in block {
// 			if let Some(err) = self.stmt(stmt) {
// 				return Some(err);
// 			}
// 		}
// 		return None;
// 	}

// 	fn stmt(&mut self, stmt: &Stmt) -> Option<FluxError> {
// 		match stmt {
// 			Stmt::If(if_) => self.if_stmt(if_),
// 			Stmt::VarDecl(var) => self.var_decl(var),
// 			Stmt::Return(ret) => self.ret_stmt(ret),
// 			Stmt::ExprStmt(expr) => self.expr(expr).1,
// 			_ => Some(self.error(
// 				format!(
// 					"internal compiler error: could not codegen statement: {:?}",
// 					stmt
// 				),
// 				FluxErrorCode::CodegenCouldNotCodegenStmt,
// 				vec![],
// 			)),
// 		}
// 	}

// 	fn if_stmt(&mut self, if_: &If) -> Option<FluxError> {
// 		let (cond, err) = self.expr(&if_.condition);
// 		if let Some(err) = err {
// 			return Some(err);
// 		}

// 		unsafe {
// 			let f = self
// 				.tmp_data
// 				.cur_fn
// 				.expect("expected to be in function: internal compiler error");
// 			let then = LLVMAppendBasicBlock(f, str_to_cstring("then"));
// 			let else_ = LLVMCreateBasicBlockInContext(self.ctx, str_to_cstring("else"));
// 			let mut cont: LLVMBasicBlockRef = ptr::null_mut();
// 			if if_.else_.is_some() {
// 				cont = LLVMCreateBasicBlockInContext(self.ctx, str_to_cstring("continue"));
// 				self.tmp_data.merging_bb = Some(cont);
// 			} else {
// 				self.tmp_data.merging_bb = Some(else_);
// 			}

// 			LLVMBuildCondBr(self.builder, cond, then, else_);

// 			self.tmp_data.cur_bb = Some(then);
// 			LLVMPositionBuilderAtEnd(self.builder, then);

// 			let name = get_basicblock_name(then);
// 			let initial_scope = self.symbol_table.cur_scope.clone();
// 			self
// 				.symbol_table
// 				.scope_conns
// 				.push((initial_scope.clone(), name.clone()));
// 			self
// 				.symbol_table
// 				.scope_values
// 				.insert(name.clone(), HashMap::new());
// 			self.symbol_table.cur_scope = name;

// 			if let Some(err) = self.block_stmt(&if_.then) {
// 				return Some(err);
// 			}
// 			if if_.else_.is_some() {
// 				LLVMBuildBr(self.builder, cont);
// 			} else {
// 				LLVMBuildBr(self.builder, else_);
// 			}

// 			self.tmp_data.cur_bb = Some(else_);
// 			LLVMPositionBuilderAtEnd(self.builder, else_);

// 			if let Some(else_block) = &if_.else_ {
// 				let name = get_basicblock_name(else_);
// 				self
// 					.symbol_table
// 					.scope_conns
// 					.push((initial_scope, name.clone()));
// 				self
// 					.symbol_table
// 					.scope_values
// 					.insert(name.clone(), HashMap::new());
// 				self.symbol_table.cur_scope = name;

// 				LLVMAppendExistingBasicBlock(f, else_);
// 				self.block_stmt(else_block);

// 				LLVMBuildBr(self.builder, cont);
// 				LLVMAppendExistingBasicBlock(f, cont);
// 				self.tmp_data.cur_bb = Some(cont);
// 				LLVMPositionBuilderAtEnd(self.builder, cont);
// 			} else {
// 				LLVMAppendExistingBasicBlock(f, else_);
// 				LLVMPositionBuilderAtEnd(self.builder, else_);
// 			}
// 		}

// 		return None;
// 	}

// 	fn ret_stmt(&mut self, ret: &Return) -> Option<FluxError> {
// 		unsafe {
// 			match &ret.val {
// 				Some(x) => {
// 					let (e, err) = self.expr(&x);
// 					if let Some(err) = err {
// 						return Some(err);
// 					}
// 					LLVMBuildRet(self.builder, e);
// 					return None;
// 				}
// 				None => {
// 					LLVMBuildRetVoid(self.builder);
// 					return None;
// 				}
// 			};
// 		}
// 	}
// }
