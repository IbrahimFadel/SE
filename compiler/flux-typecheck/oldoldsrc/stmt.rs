// use flux_ast::{Return, VarDecl};

// use super::*;

// impl<'a> TypecheckCtx<'a> {
// 	pub fn check_stmt(&mut self, stmt: &'a mut Stmt) -> Option<FluxError> {
// 		match stmt {
// 			Stmt::VarDecl(var) => self.check_var(var),
// 			Stmt::Return(ret) => self.check_ret(ret),
// 			Stmt::ExprStmt(expr) => self.check_expr(expr),
// 			_ => None,
// 		}
// 	}

// 	fn check_ret(&mut self, ret: &'a mut Return) -> Option<FluxError> {
// 		if let Some(x) = &mut ret.val {
// 			if let Some(err) = self.check_expr(x) {
// 				return Some(err);
// 			}
// 		}
// 		return None;
// 	}

// 	fn check_var(&mut self, var: &'a mut VarDecl) -> Option<FluxError> {
// 		self.expecting_ty = Some(&var.type_);
// 		for name in &var.names {
// 			self.var_types.insert(name.to_string(), &var.type_);
// 		}

// 		for val in &mut var.values {
// 			if let Some(err) = self.check_expr(val) {
// 				return Some(err);
// 			}
// 		}

// 		return None;
// 	}
// }
