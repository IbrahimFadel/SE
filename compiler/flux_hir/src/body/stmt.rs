use super::*;

impl<'a, 'pkgs> LowerCtx<'a, 'pkgs> {
    pub(crate) fn lower_stmt(
        &mut self,
        stmt: ast::Stmt,
        generic_params: &GenericParams,
    ) -> (bool, Typed<ExprIdx>) {
        match stmt {
            ast::Stmt::LetStmt(let_stmt) => (false, self.lower_let_expr(let_stmt, generic_params)),
            ast::Stmt::ExprStmt(expr_stmt) => {
                (false, self.lower_expr_stmt(expr_stmt, generic_params))
            }
            ast::Stmt::TerminatorExprStmt(terminator_expr_stmt) => (
                true,
                self.lower_terminator_expr_stmt(terminator_expr_stmt, generic_params),
            ),
        }
    }

    // Always return unit `()`
    fn lower_let_expr(
        &mut self,
        let_stmt: ast::LetStmt,
        generic_params: &GenericParams,
    ) -> Typed<ExprIdx> {
        let name = self.lower_name(let_stmt.name());
        let ty = let_stmt
            .ty()
            .map(|ty| self.lower_type(Some(ty), generic_params).inner)
            .unwrap_or_else(|| {
                self.tckh
                    .tenv
                    .insert(Type::Unknown.file_span(self.file_id, name.span))
            });
        let val = self.lower_expr(let_stmt.value(), generic_params);
        let unification_span = self.tckh.tenv.get_filespan(&val.tid);
        self.tckh
            .unify(ty, val.tid, unification_span)
            .unwrap_or_else(|err| self.diagnostics.push(err));

        let tid = let_stmt.ty().map_or(val.tid, |_| ty);

        self.tckh.tenv.insert_local(name.inner, tid);
        self.alloc_expr(Expr::Tuple(vec![])).with_type(val.tid)
    }

    fn lower_expr_stmt(
        &mut self,
        expr_stmt: ast::ExprStmt,
        generic_params: &GenericParams,
    ) -> Typed<ExprIdx> {
        self.lower_expr(expr_stmt.expr(), generic_params)
    }

    fn lower_terminator_expr_stmt(
        &mut self,
        terminator_expr_stmt: ast::TerminatorExprStmt,
        generic_params: &GenericParams,
    ) -> Typed<ExprIdx> {
        self.lower_expr(terminator_expr_stmt.expr(), generic_params)
    }
}
