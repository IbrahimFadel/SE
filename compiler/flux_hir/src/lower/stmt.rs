use crate::hir::{LetStmt, Stmt};

use super::*;

/// The lowered [`Stmt`], whether or not the [`ast::Stmt`] had a semicolon (needed for
/// typechecking), and its type
type StmtResult = (Stmt, bool, TypeId);

impl LoweringCtx {
    pub(crate) fn lower_stmt(&mut self, stmt: ast::Stmt) -> StmtResult {
        match stmt {
            ast::Stmt::LetStmt(let_stmt) => self.lower_let_stmt(let_stmt),
            ast::Stmt::ExprStmt(expr) => self.lower_expr_stmt(expr),
        }
    }

    fn lower_let_stmt(&mut self, let_stmt: ast::LetStmt) -> StmtResult {
        let name = self.lower_node(
            let_stmt.name(),
            |this, _| {
                Spanned::new(
                    this.interner.get_or_intern_static(POISONED_STRING_VALUE),
                    this.span_node(&let_stmt),
                )
            },
            |this, name| Spanned::new(name.ident().unwrap().text_key(), this.span_node(&let_stmt)),
        );

        let ty = if let Some(ty) = let_stmt.ty() {
            let ty = self.lower_type(ty);
            self.tchk.tenv.insert(self.file_spanned(self.to_ts_ty(&ty)))
        } else {
            self.tchk.tenv.insert(self.file_spanned(Spanned::new(
                ts::Type::new(TypeKind::Unknown),
                self.span_node(&let_stmt),
            )))
        };

        let (value, value_ty_id) = self.lower_node(
            let_stmt.value(),
            |this, _| {
                (
                    this.exprs
                        .alloc(Spanned::new(Expr::Error, this.span_node(&let_stmt))),
                    this.tchk.tenv.insert(this.file_spanned(Spanned::new(
                        ts::Type::new(TypeKind::Unknown),
                        this.span_node(&let_stmt),
                    ))),
                )
            },
            |this, expr| this.lower_expr(expr),
        );

        let result = self
            .tchk
            .unify(ty, value_ty_id, self.file_span(self.span_node(&let_stmt)));
        self.maybe_emit_diagnostic(result);

        self.tchk.tenv.insert_local_to_scope(name.inner, ty);

        (Stmt::LetStmt(LetStmt { name, ty, value }), true, ty)
    }

    fn lower_expr_stmt(&mut self, expr: ast::ExprStmt) -> StmtResult {
        let (idx, ty_id) = self.lower_node(
            expr.expr(),
            |this, _| {
                let expr = Spanned::new(Expr::Error, this.span_node(&expr));
                let ty_id = this.tchk.tenv.insert(
                    this.file_spanned(Spanned::new(ts::Type::new(TypeKind::Unknown), expr.span)),
                );
                let idx = this.exprs.alloc(expr);
                (idx, ty_id)
            },
            |this, expr| this.lower_expr(expr),
        );
        let has_semicolon = expr.semicolon().is_some();
        (Stmt::ExprStmt(idx), has_semicolon, ty_id)
    }
}
