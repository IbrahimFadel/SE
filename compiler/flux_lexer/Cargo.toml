[package]
name = "flux_lexer"
version = "0.1.0"
edition = "2021"

[dependencies]
logos = "0.14.0"
text-size = "1.1.1"
salsa = { path = "../../lib/salsa" }
