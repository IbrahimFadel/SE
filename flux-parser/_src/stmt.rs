use flux_ast::{BlockStmt, Expr, If, Mod, Return, Spanned, Stmt, VarDecl};
use flux_error::{PIErrorCode, Span};
use flux_lexer::token::TokenKind;

use crate::{
	expr::{expr, ident, type_expr},
	tok_val, ParseInput,
};

pub fn mod_stmt(input: &mut ParseInput, pub_: Spanned<bool>) -> Spanned<Mod> {
	let start = input.tok().span.start;
	input.next();
	let name = ident(input);
	input.expect(
		TokenKind::Semicolon,
		input.error(
			"expected `;` after mod statement".to_owned(),
			PIErrorCode::ParseExpectedSemicolonAfterModStmt,
			vec![],
		),
	);
	if input.tok().kind == TokenKind::Semicolon {
		input.next();
	}
	let end = input.tok().span.start;
	Spanned::new(Mod::new(pub_, name), Span::new(start..end, input.file_id))
}

pub fn block(input: &mut ParseInput) -> BlockStmt {
	input.expect(
		TokenKind::LBrace,
		input.error(
			"expected `{` at beginning of block".to_owned(),
			PIErrorCode::ParseExpectedLBraceInBlock,
			vec![(
				format!(
					"expected `{{` at beginning of block, instead got {}",
					tok_val(&input.program, input.tok())
				),
				Span::new(input.tok().span.clone(), input.file_id),
			)],
		),
	);
	if input.tok().kind == TokenKind::LBrace {
		input.next();
	}

	let mut stmts = vec![];
	while input.tok().kind != TokenKind::RBrace && input.tok().kind != TokenKind::EOF {
		if input.tok().kind == TokenKind::BlockComment || input.tok().kind == TokenKind::LineComment {
			input.next();
			continue;
		}
		let stmt = stmt(input);
		if stmt.node == Stmt::Error {
			break;
		}
		stmts.push(stmt);
	}

	input.expect(
		TokenKind::RBrace,
		input.error(
			"expected `}` at end of block".to_owned(),
			PIErrorCode::ParseExpectedRBraceInBlock,
			vec![(
				format!(
					"expected `}}` at end of block, instead got {}",
					tok_val(&input.program, input.tok())
				),
				Span::new(input.tok().span.clone(), input.file_id),
			)],
		),
	);
	if input.tok().kind == TokenKind::RBrace {
		input.next();
	}

	return stmts;
}

fn stmt(input: &mut ParseInput) -> Spanned<Stmt> {
	if input.tok().kind > TokenKind::TypesBegin && input.tok().kind < TokenKind::TypesEnd {
		return var_decl_stmt(input);
	} else if input
		.typenames
		.contains(&tok_val(&input.program, input.tok()))
	{
		return var_decl_stmt(input);
	} else if input.tok().kind == TokenKind::Return {
		return return_stmt(input);
	} else if input.tok().kind == TokenKind::If {
		return if_stmt(input);
	} else {
		let expr = expr(input);
		input.expect(
			TokenKind::Semicolon,
			input.error(
				"expected `;` after expression".to_owned(),
				PIErrorCode::ParseExpectedSemicolonAfterExpr,
				vec![],
			),
		);
		if input.tok().kind == TokenKind::Semicolon {
			input.next();
		}
		if expr == Spanned::new(Expr::Error, Span::new(0..0, input.file_id)) {
			return Spanned::new(Stmt::Error, Span::new(0..0, input.file_id));
		}
		return Spanned::new(
			Stmt::ExprStmt(expr.node),
			Span::new(expr.span.range.start..expr.span.range.end, input.file_id),
		);
	}
}

fn if_stmt(input: &mut ParseInput) -> Spanned<Stmt> {
	let start = input.tok().span.start;
	input.next();
	let condition = expr(input);
	let then = block(input);
	let mut else_ = None;
	if input.tok().kind == TokenKind::Else {
		input.next();
		if input.tok().kind == TokenKind::If {
			let if_ = if_stmt(input);
			else_ = Some(vec![if_]);
		} else {
			else_ = Some(block(input));
		}
	}
	let end = input.tok().span.start;
	Spanned::new(
		Stmt::If(If::new(Box::from(condition), then, else_)),
		Span::new(start..end, input.file_id),
	)
}

fn return_stmt(input: &mut ParseInput) -> Spanned<Stmt> {
	let start = input.tok().span.start;
	input.next();
	if input.tok().kind == TokenKind::Semicolon {
		input.next();
		return Spanned::new(
			Stmt::Return(Return::new(None)),
			Span::new(start..input.tok().span.start, input.file_id),
		);
	}
	let expr = expr(input);
	input.expect(
		TokenKind::Semicolon,
		input.error(
			"expected `;` after return statement".to_owned(),
			PIErrorCode::ParseExpectedSemicolonAfterReturnStmt,
			vec![],
		),
	);
	if input.tok().kind == TokenKind::Semicolon {
		input.next();
	}
	let end = input.tok().span.start;
	Spanned::new(
		Stmt::Return(Return::new(Some(expr))),
		Span::new(start..end, input.file_id),
	)
}

fn var_decl_stmt(input: &mut ParseInput) -> Spanned<Stmt> {
	let start = input.tok().span.start;
	let ty = type_expr(input);

	let mut_start = 0;
	let mut_end = 0;

	if input.tok().kind != TokenKind::Ident {
		input.errs.push(input.error(
			"expected identifier in variable declaration".to_owned(),
			PIErrorCode::ParseExpectedIdentVarDecl,
			vec![(
				format!(
					"expected identifier in variable declaration, instead got `{}`",
					tok_val(&input.program, input.tok())
				),
				Span::new(input.tok().span.clone(), input.file_id),
			)],
		));
	}

	let names_start = input.tok().span.start;
	let mut names_end = input.tok().span.end;
	let mut names = vec![];
	while input.tok().kind != TokenKind::Eq && input.tok().kind != TokenKind::Semicolon {
		names_end = input.tok().span.clone().end;
		let name = ident(input);
		names.push(name);
		if input.tok().kind != TokenKind::Eq && input.tok().kind != TokenKind::Semicolon {
			input.expect(
				TokenKind::Comma,
				input.error(
					"expected `,` in variable declaration identifier list".to_owned(),
					PIErrorCode::ParseExpectedCommaInVarDeclIdentList,
					vec![(
						format!(
							"expected `,` in variable declaration identifier list, instead got `{}`",
							tok_val(&input.program, input.tok())
						),
						Span::new(input.tok().span.clone(), input.file_id),
					)],
				),
			);
			if input.tok().kind == TokenKind::Comma {
				input.next();
			}
		}
	}

	input.expect(
		TokenKind::Eq,
		input.error(
			"expected `=` after variable declaration identifier list".to_owned(),
			PIErrorCode::ParseExpectedEqVarDeclIdentList,
			vec![(
				format!(
					"expected `=` after variable declaration identifier list, instead got `{}`",
					tok_val(&input.program, input.tok())
				),
				Span::new(input.tok().span.clone(), input.file_id),
			)],
		),
	);
	input.next();

	let values_start = input.tok().span.start;
	let mut values_end = input.tok().span.end;
	let mut values = vec![];
	while input.tok().kind != TokenKind::Semicolon {
		values_end = input.tok().span.end;
		let val = expr(input);
		values.push(val);
		if input.tok().kind == TokenKind::Comma {
			input.next();
			if input.tok().kind == TokenKind::Semicolon {
				input.errs.push(input.error(
					"expected expression after comma in variable declaration value list".to_owned(),
					PIErrorCode::ParseExpectedExprAfterCommaVarDeclValueList,
					vec![(
						format!(
								"expected expression after comma in variable declaration value list, instead got `{}`",
								tok_val(&input.program, input.tok())
							),
						Span::new(input.tok().span.clone(), input.file_id),
					)],
				));
				break;
			}
		} else if input.tok().kind != TokenKind::Semicolon {
			input.errs.push(input.error(
				"expected `;` after variable declaration".to_owned(),
				PIErrorCode::ParseExpectedSemicolonAfterVarDecl,
				vec![(
					format!(
						"expected `;` after variable declaration, instead got `{}`",
						tok_val(&input.program, input.tok())
					),
					Span::new(input.tok().span.clone(), input.file_id),
				)],
			));
			break;
		}
	}
	if input.tok().kind == TokenKind::Semicolon {
		input.next();
	}

	if names.len() > 1 && values.len() > 1 && values.len() != names.len() {
		if names.len() > values.len() {
			input.errs.push(input.error("more variables than values in variable declaration".to_owned(), PIErrorCode::ParseMoreIdentsThanValsVarDecl, vec![
				(format!("more variables than values in variable declaration: {} values assigned to {} variables", values.len(), names.len()), Span::new(names_start..input.tok().span.end, input.file_id)),
				(format!("{} variables declared", names.len()), Span::new(names_start..names_end, input.file_id)),
				(format!("{} values assigned", values.len()), Span::new(values_start..values_end, input.file_id)),
				("(hint) you can assign one value to multiple variables".to_owned(), Span::new(names_start..input.tok().span.end, input.file_id)),
			]));
		} else {
			input.errs.push(input.error("more values than variables in variable declaration".to_owned(), PIErrorCode::ParseMoreValsThanIdentsVarDecl, vec![
				(format!("more values than variables in variable declaration: {} values assigned to {} variables", values.len(), names.len()), Span::new(names_start..input.tok().span.end, input.file_id)),
				(format!("{} variables declared", names.len()), Span::new(names_start..names_end, input.file_id)),
				(format!("{} values assigned", values.len()), Span::new(values_start..values_end, input.file_id)),
			]));
		}
	}

	let end = input.tok().span.start;
	return Spanned::new(
		Stmt::VarDecl(VarDecl::new(
			Spanned::new(false, Span::new(mut_start..mut_end, input.file_id)),
			ty,
			names,
			values,
		)),
		Span::new(start..end, input.file_id),
	);
}
