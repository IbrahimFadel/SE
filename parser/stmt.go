package parser

import (
	"fmt"

	"github.com/IbrahimFadel/pi-lang/ast"
)

func (p *Parser) ParseBlockStmt() (ast.BlockStmt, error) {
	var block ast.BlockStmt

	p.Expect(ast.TokenTypeOpenCurlyBracket, "expected '{' at beggining of block statement")
	startPos := p.CurTok.Pos
	p.EatToken()

	for p.CurTok.TokenType != ast.TokenTypeCloseCurlyBracket {
		stmt, err := p.ParseStatement()
		if err != nil {
			return block, fmt.Errorf("could not parse statement: %s", err.Error())
		}
		block.List = append(block.List, stmt)
	}

	endPos := p.CurTok.Pos
	p.EatToken()

	block.Start = startPos
	block.End = endPos

	return block, nil
}

func (p *Parser) ParseStatement() (ast.Stmt, error) {
	var stmt ast.Stmt

	switch p.CurTok.TokenType {
	default:
		return stmt, fmt.Errorf("no method for parsing statement: %s", p.CurTok.Value)
	case ast.TokenTypeReturn:
		return p.ParseReturn()
	case ast.TokenTypeMut, ast.TokenTypeConst:
		return p.ParseVarDecl()
	case ast.TokenTypeIdentifier:
		return p.ParseExpr()
	}
}

func (p *Parser) ParseReturn() (ast.ReturnStmt, error) {
	var ret ast.ReturnStmt

	retPos := p.CurTok.Pos

	p.EatToken()

	ret.Type = p.CurFunc.FuncType.Return
	p.CurType = ret.Type
	retVal, err := p.ParseExpr()
	if err != nil {
		return ret, fmt.Errorf("could not parse expression: %s", err.Error())
	}
	ret.Value = retVal
	ret.ReturnPos = retPos

	return ret, nil
}
