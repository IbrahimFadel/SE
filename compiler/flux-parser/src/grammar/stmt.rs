use flux_lexer::T;

use super::{expr::type_expr, *};

pub(crate) fn stmt(p: &mut Parser) -> CompletedMarker {
	if p.at(T!(let))
		|| p.at(T!(iN))
		|| p.at(T!(uN))
		|| p.at(T!(f32))
		|| p.at(T!(f64))
		|| (p.at(T!(ident)) && matches!(p.peek_next(), Some(T!(ident)) | Some(T!(eq))))
		|| p.at(T!(lparen))
	{
		var_decl(p)
	} else if p.at(T!(return)) {
		return_stmt(p)
	} else {
		let m = p.start();
		expr::expr(p, true);
		if p.at(T!(semicolon)) {
			p.bump();
		}
		m.complete(p, SyntaxKind::ExprStmt)
	}
}

fn return_stmt(p: &mut Parser) -> CompletedMarker {
	let m = p.start();
	p.bump();
	if p.at(T!(semicolon)) {
		p.bump();
		return m.complete(p, SyntaxKind::ReturnStmt);
	}
	expr::expr(p, true);
	p.expect(
		T!(semicolon),
		format!("expected `;` after return statement"),
	);
	m.complete(p, SyntaxKind::ReturnStmt)
}

// fn else_if_stmt(p: &mut Parser) -> CompletedMarker {
// 	let m = p.start();
// 	p.expect(T!(if), format!("expected `if` in else if statement"));
// 	expr::expr(p, false);
// 	block(p);
// 	m.complete(p, SyntaxKind::IfStmt)
// }

fn var_decl(p: &mut Parser) -> CompletedMarker {
	let m = p.start();
	if p.at(T!(let)) {
		p.bump();
	} else {
		type_expr(p);
	}
	p.expect(
		TokenKind::Ident,
		format!("expected identifier in variable declaration"),
	);
	p.expect(
		TokenKind::Eq,
		format!("expected `=` in variable declaration"),
	);
	expr::expr(p, true);
	p.expect(
		TokenKind::SemiColon,
		format!("expected `;` after variable declaration"),
	);
	m.complete(p, SyntaxKind::VarDecl)
}

#[cfg(test)]
mod tests {
	use crate::test_stmt_str;
	test_stmt_str!(var_decl, "i32 x = 1;");
	test_stmt_str!(
		var_decl_fcont,
		r#"i32 y =
	i32 x = 1;"#
	);
}
