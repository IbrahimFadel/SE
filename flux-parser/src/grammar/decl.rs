use super::{expr::type_expr, *};
use crate::grammar::stmt::block;
use flux_lexer::T;

pub(crate) fn top_level_decl(p: &mut Parser) {
	if p.at(T!(pub)) {
		pub_tob_level_decl(p);
	} else if p.at(T!(mod)) {
		mod_decl(p);
	} else if p.at(T!(use)) {
		use_decl(p);
	} else if p.at(T!(fn)) {
		fn_decl(p);
	} else if p.at(T!(type)) {
		type_decl(p);
	} else {
		if !p.at(T!(comment)) {
			p.error(format!("expected top level declaration"));
		}
	}
}

fn pub_tob_level_decl(p: &mut Parser) {
	if p.next_at(T!(fn)) {
		fn_decl(p);
	} else if p.next_at(T!(type)) {
		type_decl(p);
	} else {
		if !p.at(T!(comment)) {
			p.error(format!("expected top level declaration"));
		}
	}
}

fn mod_decl(p: &mut Parser) -> CompletedMarker {
	assert!(p.at(T!(mod)));
	let m = p.start();
	p.bump();
	p.expect(T!(ident), format!("expected identifier in mod declaration"));
	p.expect(T!(semicolon), format!("expected `;` after mod declaration"));
	m.complete(p, SyntaxKind::ModDecl)
}

fn use_decl(p: &mut Parser) -> CompletedMarker {
	assert!(p.at(T!(use)));
	let m = p.start();
	p.bump();

	while !p.at(T!(semicolon)) && !p.at_end() {
		p.expect(T!(ident), format!("expected identifier in `use` path"));
		if p.at(T!(semicolon)) {
			break;
		} else {
			if p.at(T!(::)) {
				p.bump();
			} else {
				p.error(format!("expected `::` or `;` in `use` path"));
				break;
			}
		}
	}

	p.expect(T!(semicolon), format!("expected `;` after `use` path"));
	m.complete(p, SyntaxKind::UseDecl)
}

fn type_decl(p: &mut Parser) -> CompletedMarker {
	assert!(p.at(T!(type)) || p.at(T!(pub)));
	let m = p.start();
	if p.at(T!(pub)) {
		p.bump()
	}
	p.expect(T!(type), format!("expected `type` in type declaration"));
	p.expect(
		T!(ident),
		format!("expected identifier in type declaration"),
	);
	type_expr(p);
	m.complete(p, SyntaxKind::TypeDecl)
}

fn fn_decl(p: &mut Parser) -> CompletedMarker {
	assert!(p.at(T!(fn)) || p.at(T!(pub)));
	let m = p.start();
	if p.at(T!(pub)) {
		p.bump()
	}
	p.expect(T!(fn), format!("expected `fn` in function declaration"));
	p.expect(
		T![ident],
		format!("expected identifier in function declaration"),
	);
	if p.at(T![<]) {
		generic_list(p);
	}
	fn_params(p);
	if p.at(T!(->)) {
		p.bump();
		type_expr(p);
	}
	block(p);
	m.complete(p, SyntaxKind::FnDecl)
}

fn generic_list(p: &mut Parser) -> CompletedMarker {
	assert!(p.at(T![<]));
	let m = p.start();
	p.expect(
		T![ident],
		format!("expected identifier in function generic list"),
	);
	m.complete(p, SyntaxKind::GenericList)
}

fn fn_params(p: &mut Parser) {
	p.expect(
		T![lparen],
		format!("expected `(` at beginning of function parameter list"),
	);
	while !p.at(T![rparen]) && !p.at_end() {
		fn_param(p);
		if !p.at(T![rparen]) {
			if !p.at(T![comma]) {
				p.error(format!(
					"expected either `,` or `)` in function parameter list"
				));
				break;
			} else {
				p.bump();
			}
		}
	}
	p.expect(
		T![rparen],
		format!("expected `)` at end of function parameter list"),
	);
}

fn fn_param(p: &mut Parser) -> CompletedMarker {
	let m = p.start();
	if p.at(T![mut]) {
		p.bump();
	}
	type_expr(p);
	p.expect(
		T![ident],
		format!("expected identifier in function parameter"),
	);
	m.complete(p, SyntaxKind::FnParam)
}

#[cfg(test)]
mod test {
	use crate::test_decl_str;

	test_decl_str!(fn_decl, r#"fn main(i15 argc, mut u4 test) -> f32 {}"#);
	test_decl_str!(fn_decl_ret_void, r#"fn main(i15 argc, mut u4 test) {}"#);
	test_decl_str!(ty_decl_prim, r#"type Foo i32"#);
	test_decl_str!(
		ty_decl_struct,
		r#"type Foo struct {
		i7 x;
		pub u1 y;
		mut f64 z;
		pub mut Bar a;
	}"#
	);
	test_decl_str!(
		ty_decl_interface,
		r#"type Foo interface {
			fn foo();
			pub fn bar(i32 x) -> u5;
			fn bazz(mut f64 x) -> f32;
	}"#
	);
	test_decl_str!(
		ty_decl_interface_method_missing_ret_ty,
		r#"type Foo interface {
			fn foo() -> ;
	}"#
	);
}
