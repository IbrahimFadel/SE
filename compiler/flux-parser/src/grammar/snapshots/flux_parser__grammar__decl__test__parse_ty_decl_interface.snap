---
source: flux-parser/src/grammar/decl.rs
expression: "&parse.debug_tree()"
---
TypeDecl@0..94
  TypeKw@0..4 "type"
  Whitespace@4..5 " "
  Ident@5..8 "Foo"
  Whitespace@8..9 " "
  InterfaceType@9..94
    InterfaceKw@9..18 "interface"
    Whitespace@18..19 " "
    RBrace@19..20 "{"
    Whitespace@20..24 "\n\t\t\t"
    InterfaceMethod@24..37
      FnKw@24..26 "fn"
      Whitespace@26..27 " "
      Ident@27..30 "foo"
      LParen@30..31 "("
      RParen@31..32 ")"
      SemiColon@32..33 ";"
      Whitespace@33..37 "\n\t\t\t"
    InterfaceMethod@37..65
      PubKw@37..40 "pub"
      Whitespace@40..41 " "
      FnKw@41..43 "fn"
      Whitespace@43..44 " "
      Ident@44..47 "bar"
      LParen@47..48 "("
      FnParam@48..53
        PrimitiveType@48..52
          INKw@48..51 "i32"
          Whitespace@51..52 " "
        Ident@52..53 "x"
      RParen@53..54 ")"
      Whitespace@54..55 " "
      Arrow@55..57 "->"
      Whitespace@57..58 " "
      PrimitiveType@58..60
        UNKw@58..60 "u5"
      SemiColon@60..61 ";"
      Whitespace@61..65 "\n\t\t\t"
    InterfaceMethod@65..93
      FnKw@65..67 "fn"
      Whitespace@67..68 " "
      Ident@68..72 "bazz"
      LParen@72..73 "("
      FnParam@73..82
        MutKw@73..76 "mut"
        Whitespace@76..77 " "
        PrimitiveType@77..81
          F64Kw@77..80 "f64"
          Whitespace@80..81 " "
        Ident@81..82 "x"
      RParen@82..83 ")"
      Whitespace@83..84 " "
      Arrow@84..86 "->"
      Whitespace@86..87 " "
      PrimitiveType@87..90
        F32Kw@87..90 "f32"
      SemiColon@90..91 ";"
      Whitespace@91..93 "\n\t"
    RBrace@93..94 "}"
